source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -418,23 +418,38 @@ std::vector<uint8_t> ContextImpl::parseAlpnProtocols(const std::string& alpn_pro
   return out;
 }
 
-bssl::UniquePtr<SSL> ContextImpl::newSsl(const Network::TransportSocketOptions*) {
+bssl::UniquePtr<SSL>
+ContextImpl::newSsl(const Network::TransportSocketOptionsConstSharedPtr& options) {
   // We use the first certificate for a new SSL object, later in the
   // SSL_CTX_set_select_certificate_cb() callback following ClientHello, we replace with the
   // selected certificate via SSL_set_SSL_CTX().
-  return bssl::UniquePtr<SSL>(SSL_new(tls_contexts_[0].ssl_ctx_.get()));
+  auto ssl_con = bssl::UniquePtr<SSL>(SSL_new(tls_contexts_[0].ssl_ctx_.get()));
+  if (tls_async_cert_validation_) {
+    SSL_set_app_data(ssl_con.get(), &options);
+  }
+  return ssl_con;
 }
 
 int ContextImpl::verifyCallback(X509_STORE_CTX* store_ctx, void* arg) {
   ContextImpl* impl = reinterpret_cast<ContextImpl*>(arg);
   SSL* ssl = reinterpret_cast<SSL*>(
       X509_STORE_CTX_get_ex_data(store_ctx, SSL_get_ex_data_X509_STORE_CTX_idx()));
   auto cert = bssl::UniquePtr<X509>(SSL_get_peer_certificate(ssl));
+  const Network::TransportSocketOptions* transport_socket_options = nullptr;
+  if (impl->tls_async_cert_validation_) {
+    auto transport_socket_options_shared_ptr_ptr =
+        static_cast<const Network::TransportSocketOptionsConstSharedPtr*>(SSL_get_app_data(ssl));
+    ASSERT(transport_socket_options_shared_ptr_ptr);
+    transport_socket_options = (*transport_socket_options_shared_ptr_ptr).get();
```

Comment:
Is the only difference between these two branches the ASSERT()? I can't quite seem to convince myself of it.

Reply:
Hm. So line 441 calls `SSL_get_app_data(ssl)` and casts it to TransportSocketOptionsConstSharedPtr*. but link 446 calls the same method and casts it to TransportSocketOptions*. This is confusing to me. What is the actual object stored in SSL_get_app_data? Maybe the type of the object is different in the old/new codepath?

Reply:
Right, in the old path SSL_set_app_data() takes the TransportSocketOptions* which is from TransportSocketOptionsConstSharedPtr, but the in the new path SSL_set_app_data() takes a pointer to TransportSocketOptionsConstSharedPtr itself.

Reply:
Ah, I see. Thanks for the explanation!

Storing a pointer to a shared pointer sets off my spidey sense a bit. A shared pointer implies the that object is owned by whoever has the last shared pointer to it, whereas a pointer implies that the object is owned elsewhere. But I guess SSL_set_app_data() requires a raw pointer and we can't put a smart pointer in there, right? And I guess this is safe because we require the caller to own Options and make it outlive the context, right?

Reply:
Oh, I see from the PR description that you want the shared_ptr so that you can extend the lifetime of the object via what's stored in the `SSL`. Can you elaborate on the use case for this?

Given that this PR doesn't yet have any external behavior change, it might simplify to remove the runtime guard.

Reply:
Okay, reverted runtime guard. PTAL!